libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_DNS.c

```diff
@@ -1490,7 +1494,18 @@ BaseType_t xReturn;
 					/* Confirm that the record is still fresh. */
 					if( ulCurrentTimeSeconds < ( xDNSCache[ x ].ulTimeWhenAddedInSeconds + FreeRTOS_ntohl( xDNSCache[ x ].ulTTL ) ) )
 					{
-						*pulIP = xDNSCache[ x ].ulIPAddress;
+#if( ipconfigDNS_CACHE_ADDRESSES_PER_ENTRY > 1 )
+                        /* The ucCurrentIPAddress value increments without bound and will rollover, */
+                        /*  modulo it by the number of IP addresses to keep it in range.     */
+                        /*  Also perform a final modulo by the max number of IP addresses    */
+                        /*  per DNS cache entry to prevent out-of-bounds access in the event */
+                        /*  that ucNumIPAddresses has been corrupted.                        */
+                        ulIPAddressIndex = ( xDNSCache[ x ].ucCurrentIPAddress %
+                                        xDNSCache[ x ].ucNumIPAddresses ) % ipconfigDNS_CACHE_ADDRESSES_PER_ENTRY;
+
+                        xDNSCache[ x ].ucCurrentIPAddress++;
```

Comment:
Suppose that 10 addresses were found. When calling this function repeatedly, I would see the addresses 0..9, 0..9, and when `ucCurrentIPAddress` becomes 255, I would see 0..5, followed by 0..9 again?

Why not:
~~~c
if( ++xDNSCache[ x ].ucCurrentIPAddress == xDNSCache[ x ].ucNumIPAddresses )
{
	xDNSCache[ x ].ucCurrentIPAddress = 0;
}
~~~

Making `ucCurrentIPAddress` a `uint8_t` is no problem, DNS can not return more than +/- 27 answers.

Reply:
One more general question about the change:
Suppose I'm looking up the addresses of some host and it receives 5 IP-addresses.
I will call `FreeRTOS_gethostbyname()` repeatedly to see all addresses.
How do I know when I've seen all IP-addresses? Should I remember the first IP-address received?
Why not return 0 after the last IP-address was returned? And then the first IP-address again?

Reply:
Applications using `FreeRTOS_gethostbyname()` should interpret the result as an independent scalar value rather than a member of a list, and they should _not_ prefer any value over any other.  When we have an implementation of `FreeRTOS_getaddrinfo()`, applications will have the option of choosing a particular address, which could be useful in some circumstances.

The reasoning behind this change is to give app developers an easy way to increase reliability of their app when connecting to a load-balanced service, without requiring them to change any application logic.  As just one example, the amazon-freertos repo has dozens of tests and demos which can benefit immediately without any modification.

Reply:
> Suppose that 10 addresses were found. When calling this function repeatedly, I would see the addresses 0..9, 0..9, and when `ucCurrentIPAddress` becomes 255, I would see 0..5, followed by 0..9 again?
> 
> Why not:
> 
> ```c
> if( ++xDNSCache[ x ].ucCurrentIPAddress == xDNSCache[ x ].ucNumIPAddresses )
> {
> 	xDNSCache[ x ].ucCurrentIPAddress = 0;
> }
> ```
> 
> Making `ucCurrentIPAddress` a `uint8_t` is no problem, DNS can not return more than +/- 27 answers.

I didn't choose this option because I didn't want to have to do anything for thread safety for cases where two tasks both call `FreeRTOS_gethostbyname()`.  In the event that there was a context switch after the `if` but before the statement in `{...}`, we could have a [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) error resulting in an out-of-bound array access.

The side effect of doing the modulo approach as you have pointed out is that you the order that addresses appear may not reflect how they are stored internally, but this isn't important since the caller should make no assumption about these values.  The only thing that is promised is that `FreeRTOS_gethostbyname()` will return an answer if one is available.

Reply:
> I didn't choose this option because I didn't want to have to do anything for thread safety for cases where two tasks both call `FreeRTOS_gethostbyname()`. In the event that there was a context switch after the `if` but before the statement in `{...}`, we could have a [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) error resulting in an out-of-bound array access.

What if we do 
```
if( ++xDNSCache[ x ].ucCurrentIPAddress >= xDNSCache[ x ].ucNumIPAddresses )
{   /* Use >=  instead of == */
     .... 
}
```

Would it remove TOCTOU error? I could not think of a case where it fails. Do correct me if I am wrong.
I am just trying to reduce computation as much as we can. Modulo operation is quite expensive than a comparison statement in a lot of architectures.

Reply:
Remember that if [interrupts](https://en.wikipedia.org/wiki/Interrupt) aren't masked, a [context switch](https://en.wikipedia.org/wiki/Context_switch) can happen at any time.  And since this code is not protected by any mutual-exclusion mechanism such as [a mutex or semaphore](https://www.freertos.org/a00113.html) (and we don't want to add those if we don't have to), you can still have the error.

In any case, the code as written is simpler and avoids any possibility of this occurring.